spotbugs/src/main/java/edu/umd/cs/findbugs/detect/MultipleInstantiationsOfSingletons.java

```diff
@@ -0,0 +1,225 @@
+            methods.put(Methods.CLONE, getXMethod());
+            hasCloneMethod = true;
+        }
+
+        if (Const.CONSTRUCTOR_NAME.equals(getMethod().getName())) {
+            constructors.add(getXMethod());
+        }
+
+        super.visit(obj);
+    }
+
+    @Override
+    public boolean beforeOpcode(int seen) {
+        if (seen == Const.ATHROW && "clone".equals(getMethod().getName())) {
+            return true;
+        }
+        if (seen == Const.ARETURN && Const.CONSTRUCTOR_NAME.equals(getMethod().getName())) {
+            return true;
+        }
+        if (seen == Const.ARETURN || seen == Const.MONITORENTER) {
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public void sawOpcode(int seen) {
+        if (seen == Const.ATHROW) {
+            OpcodeStack.Item item = stack.getStackItem(0);
+            if (item != null) {
+                if (item.getSignature().equals("Ljava/lang/CloneNotSupportedException;") && cloneOnlyThrowsException) {
+                    cloneOnlyThrowsCloneNotSupportedException = true;
+                }
+            }
+        } else if (seen == Const.ARETURN) {
+            OpcodeStack.Item item = stack.getStackItem(0);
+            XField field = item.getXField();
+            if (field != null) {
+                String className = "L" + getClassName() + ";";
+                if (field.isPrivate() && field.isStatic() && field.getSignature().equals(className)) {
+                    isSingleton = true;
+                    isGetterMethodSynchronized = getMethod().isSynchronized();
+                    methods.put(Methods.INSTANCE_GETTER, getXMethod());
+                }
+            }
+        } else if (seen == Const.MONITORENTER) {
+            methodsUsingMonitor.add(getXMethod());
+        }
+    }
+
+    @Override
+    public void visitAfter(JavaClass obj) {
+        if (!isSingleton) {
+            return;
+        }
+
+        boolean hasNonPrivateConstructor = false;
+        for (XMethod constructor : constructors) {
+            if (!constructor.isPrivate()) {
+                hasNonPrivateConstructor = true;
+                methods.put(Methods.CONSTRUCTOR, constructor);
+                break;
+            }
+        }
+
+        boolean isGetterMethodUsingMonitor = methodsUsingMonitor.contains(methods.get(Methods.INSTANCE_GETTER));
+
+        if (hasNonPrivateConstructor) {
+            bugReporter.reportBug(new BugInstance(this, "SING_SINGLETON_HAS_NONPRIVATE_CONSTRUCTOR", NORMAL_PRIORITY).addClass(this).addMethod(
+                    methods.get(Methods.CONSTRUCTOR)));
+        }
+
+        if (!isGetterMethodSynchronized && !isGetterMethodUsingMonitor) {
+            bugReporter.reportBug(new BugInstance(this, "SING_SINGLETON_GETTER_NOT_SYNCHRONIZED", NORMAL_PRIORITY).addClass(this).addMethod(
+                    methods.get(Methods.INSTANCE_GETTER)));
```

Comment:
[This case is acceptable if the `clone()` method is override](https://wiki.sei.cmu.edu/confluence/display/java/MSC07-J.+Prevent+multiple+instantiations+of+singleton+objects#MSC07J.Preventmultipleinstantiationsofsingletonobjects-CompliantSolution(Overrideclone()Method)).
Please add a test case which uses the example class in that doc:

```
class MySingleton implements Cloneable {
  private static MySingleton instance;
 
  private MySingleton() {
    // Private constructor prevents instantiation by untrusted callers
  }
 
  // Lazy initialization
  public static synchronized MySingleton getInstance() {
    if (instance == null) {
      instance = new MySingleton();
    }
    return instance;
  }
 
  public Object clone() throws CloneNotSupportedException {
    throw new CloneNotSupportedException();
  }
}
```